Assets/_Dev/Runtime/Spawner.cs

```diff
@@ -8,62 +8,123 @@ namespace Playground
+        private WaveDefinition[] waves;
         [SerializeField, Tooltip("The radius around the spawner to spawn enemies.")]
         internal float spawnRadius = 5f;
-        [SerializeField, Tooltip("The duration of each spawn wave in seconds.")]
-        private float waveDuration = 10f;
         [SerializeField, Tooltip("The duration of the wait between each spawn wave in seconds.")]
         private float waveWait = 5f;
+        [SerializeField, Tooltip("When defined waves are exhausted, generate more?.")]
+        private bool generateWaves = true;
 
         [Header("Events")]
         [SerializeField, Tooltip("The event to trigger when this spawner is destroyed.")]
         public UnityEvent onDestroyed;
+        [SerializeField, Tooltip("The event to trigger when all waves are complete.")]
+        public UnityEvent onWavesComplete;
 
         List<BasicEnemyController> spawnedEnemies = new List<BasicEnemyController>();
 
+        private int currentWaveIndex = -1;
+
+        private WaveDefinition currentWave;
+
         private void Start()
         {
-            StartCoroutine(SpawnEnemies());
+            StartCoroutine(SpawnWaves());
+        }
+
+        private void NextWave()
+        {
+            currentWaveIndex++;
+            if (currentWaveIndex >= waves.Length)
+            {
+                if (!generateWaves)
+                {
+                    Debug.LogWarning("No more waves to spawn.");
+                    currentWave = null;
+                    onWavesComplete?.Invoke();
+                    StopCoroutine(SpawnWaves());
+                    return;
+                }
+                currentWave = GenerateNewWave();
+                return;
+            }
+            Debug.Log($"Starting wave {currentWaveIndex + 1} of {waves.Length}...");
+            currentWave = waves[currentWaveIndex];
+            currentWave.Reset();
+        }
+
+        private WaveDefinition GenerateNewWave()
+        {
+            Debug.Log("Generating new wave...");
+            WaveDefinition newWave = ScriptableObject.CreateInstance<WaveDefinition>();
+            // populate newWave with random values based loosely on the previous wave
+            var lastWave = currentWave;
+            if (lastWave == null)
+            {
+                lastWave = waves[waves.Length - 1];
+            }
+            // collect all enemy prefabs from all waves and then randomly select from them
+            List<BasicEnemyController> enemyPrefabs = new List<BasicEnemyController>();
+            for (int i = 0; i < waves.Length; i++)
+            {
+                enemyPrefabs.AddRange(waves[i].EnemyPrefabs);
+            }
+            newWave.Init(
+                enemyPrefabs.ToArray(),
+                Mathf.Max(lastWave.SpawnRate + Random.Range(-1f, 1f), 1f), // TODO: make this change more interesting?
+                Mathf.Max(lastWave.WaveDuration + Random.Range(-1f, 1f), 1f),
```

Comment:
This means a wave will be max 1 duration? Seems short. 

It can also be 0.

Reply:
I was trying to say that the minimum duration would be 1